sigs/jvm/rfcs/20191023-tensorflow-keras.md

```diff
@@ -0,0 +1,342 @@
+            Options options;
+
+            public Builder() {
+                options = new Options()
+            }
+
+            public Builder setArg1(Arg1 arg1) {
+                options.arg1 = arg1;
+                return this;
+            }
+            public void setArg2(Arg2 arg2) {
+                options.arg2 = arg2;
+                return this;
+            }
+
+            public Options build() {
+                return options;
+            }
+        }
+    }
+}
+```
+
+### Comparing Java API to Python
+#### Original Python MNIST
+```python
+import tensorflow as tf
+
+model = tf.keras.models.Sequential([
+  tf.keras.layers.Flatten(input_shape=(28, 28)),
+  tf.keras.layers.Dense(128, activation='relu', kernel_initializer="random_normal", bias_initializer="zeros"),
+  tf.keras.layers.Dense(10, activation='softmax', kernel_initializer="random_normal", bias_initializer="zeros")
+])
+
+model.compile(optimizer='sgd', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
+
+(X_train, y_train), (X_val, y_val) = tf.keras.datasets.load_mnist()
+model.fit(X_train, y_train, val_data=(X_val, y_val), epochs=10, batch_size=100)
+```
+
+#### MNIST using Java API
+
+```java
+public class MNISTKeras {
+    private static Model<Float> model;
+
+    static {
+        model = Sequential.of(
+            Float.class, // dtype which will be passed through layer.build
+            Layers.input(28, 28),
+            Layers.flatten(),
+            Layers.dense(128, Activations.softmax, Initializers.randomNormal, Initializers.zeros),
+            Layers.dense(10, Activations.softmax, Initializers.randomNormal, Initializers.zeros)
+        );
+    }
+
+    public static Model<Float> train(Model<Float> model) throws Exception {
+        try (Graph graph = new Graph()) {
+            Ops tf = Ops.create(graph);
+            model.compile(tf, Optimizers.sgd, Losses.sparseCategoricalCrossentropy, Metrics.accuracy);
+
+            Pair<GraphLoader<Float>, GraphLoader<Float>> loaders = MNIST.graphLoaders2D();
+            // GraphLoader objects contain AutoCloseable `Tensor` objects.
+            try (GraphLoader<Float> train = loaders.first(); GraphLoader<Float> test = loaders.second()) {
+                model.fit(tf, train, test, 10, 100);
+            }
+        }
+
+        return model;
+    }
+
+    public static void main(String[] args) throws Exception {
+        train(model);
+    }
+}
+
+```
+
+#### MNIST using Scala wrapper around Java API
+```scala
+object MNISTKeras {
+
+  val model: Model[JFloat] = Sequential.of[JFloat](
+    classOf[JFloat],
+    input(28, 28),
+    flatten(),
+    dense(128, activation = relu, kernelInitializer = randomNormal, biasInitializer = zeros),
+    dense(10, activation = softmax, kernelInitializer = randomNormal, biasInitializer = zeros)
+  )
+
+  def train(model: Model[JFloat]): Model[JFloat] = {
+      Using.resource(new Graph()) { graph => {
+        implicit val tf: Ops = Ops.create(graph)
```

Comment:
I see, `tf` is implicit with Scala, but afaik it is not possible to achieve this in any other JVM languages, so we should probably optimize the API for the most common case.